### ISSUE TITLE
Support libc++

### ISSUE THREAD

**Issue by E5ten** [falconindy/auracle#41]
One problem with libc++ support is that auracle uses C++17 features, but also features that have been removed from C++17, and by default libc++ does not allow use of features removed from C++17 while the C++ standard is set to 17, so `_LIBCPP_ENABLE_CXX17_REMOVED_FEATURES` needs to be defined. 

In addition, auracle hardcodes use of the `libstdc++fs` filesystem library, while libc++ has the `libc++fs` library, but when I try to manually replace `stdc++fs` with `c++fs` in the meson.build, while it does successfully build, auracle with segfaults upon trying to perform any operation.

**Comment by E5ten** [comment]
Actually re-enabling all of the removed C++17 features with `_LIBCPP_ENABLE_CXX17_REMOVED_FEATURES` shouldn't be necessary, defining just `_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR` should be sufficient.

**Comment by falconindy** [comment]
What removed features?

**Comment by E5ten** [comment]
ptr_fun, deprecated in C++11 and removed in C++17. I was wrong in the last comment about not needing `_LIBCPP_ENABLE_CXX17_REMOVED_FEATURES` I don't know what I did wrong in my test before that led me to believe only `_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR` would be necessary.

**Comment by falconindy** [comment]
As far as the segfault, how are you building auracle? I would imagine that a combination of libstdc++ (for c++ standard features) and libc++ (for libc++fs) could be ABI incompatible and lead to segfaults.

**Comment by E5ten** [comment]
I'm using clang with -stdlib=libc++ so it shouldn't be using libstdc++ at all.

**Comment by E5ten** [comment]
Oh sorry I forgot one other change I had to make, I had to add `#include <unordered_map>` to aur.hh, so I guess some kind of ifdef to check if libc++ is being used and including that if it is would also be necessary. I'm not sure what would be checked to be defined to see if libc++ is being used though.

**Comment by falconindy** [comment]
Apparently libstdc++ still allows string_view construction with a nullptr, but libc++ has moved on.

Tests don't seem to compile with libc++, but that's not something I'm going to be fixing myself. I'll wait on Google to get their act together (probably with a new release when they get around to appeasing the outside masses, as they internally support libc++).

**Comment by E5ten** [comment]
So then the only remaining issue would be that the meson.build still requires editing to use libc++fs instead of libstdc++fs.

**Comment by E5ten** [comment]
Actually I still get segfaults when I use certain flags, it seems the flags that cause this to happen are any which can't use an `=`, so `auracle search test --searchby maintainer` won't segfault, but `auracle search test --quiet` will. I tested this with a lot of the flags and it seems consistent, if a flag has a second argument it will not cause a segfault, but argumentless flags cause segfaults.

**Comment by falconindy** [comment]
Makes sense. Just other places where a string_view is constructed with a nullptr. string_view(nullptr) is actually undefined and various c++ libraries behave differently. https://github.com/abseil/abseil-cpp/issues/73#issuecomment-355620727 is an interesting take on things (Titus is on the c++ standards committee).

This is the kind of thing that would be easy to catch (and prevent in the future) if the tests were able to be built with libc++. I don't have time to support something that I have to manually test (as an alternative build configuration no less). It's hard for me to actually commit to maintaining this build. Given that Arch's default c++ stdlib is still provided by gcc, I have even less motivation to support this.

Additionally, supporting c++fs in the buildsystem is a minefield. I can add a flag to select a stdlib implementation, but that's only half the puzzle. There's at least three workable compilation configurations here:

1) g++, libstdc++, stdc++fs
2) clang++, libstdc++, stdc++fs
3) clang++, libc++, c++fs

Things that don't work:
1) g++, libc++, c++fs
2) g++, libc++, stdc++fs

The workable setups aren't easy to assert without significantly complicating the build system, and with minimal gain.

Related: https://github.com/mesonbuild/meson/issues/3685

**Comment by falconindy** [comment]
Ah, tests don't build because arch distributes shared libs for gmock and gtest. They're built with libstdc++ and thus have a different ABI. So, symbols aren't found when compiling with libc++.

**Comment by E5ten** [comment]
Ok, that's understandable, I can see why you don't want to add a configuration option for supporting libc++/libc++fs when the arch provided gmock and gtest don't allow for the automated testing to work. Would supporting them to the extent that if someone (like me) reported an issue that arose with libc++ such as the current segfault due to a string_view being constructed with a nullptr be off the table as well?

**Comment by falconindy** [comment]
Sure, I can fix things that are questionable according to the standard.

For any of these things, it'd be nice to have a clang-tidy check to ensure that these don't creep back in.

**Comment by E5ten** [comment]
I don't think the replacement for the environment check for debugging purposes is working as intended, because now with no change to my environment auracle always outputs debug connection info.

**Comment by falconindy** [comment]
Yep you're right, fixed.

**Comment by E5ten** [comment]
Since support within the build system is not within scope and all current problems seem to be resolved I'll close this and make a new issue should something change that causes problems with libc++, thanks!

**Comment by E5ten** [comment]
Since LLVM 9 libc++'s filesystem stuff is part of libc++ instead of in a separate library, meaning that in terms of build system support there wouldn't be that much complexity, because instead of having to choose a fs lib to use with the C++ lib you should just be able to link libstdc++fs if libstdc++ is used, and not if libc++ is, would that difference be enough to warrant libc++ support or would that still be too much complexity for the build system?

**Comment by falconindy** [comment]
I guess I'm just not sure why it should be a package-level decision (as opposed to a distro/system-level decision) as to which libc is used.

**Comment by E5ten** [comment]
At the moment isn't it even more a package-level decision because unpatched the build system forces libstdc++ instead of allowing either? The reason it would have to be implemented build system level is because of the filesystem lib, if it was just about choosing with C++ lib to use that'd get determined by the chosen -stdlib=value in CXXFLAGS.

**Comment by falconindy** [comment]
Is there currently any advantage to using libc++ over stdlibc++?

**Comment by falconindy** [comment]
So, I just built and installed libc++. I compiled auracle as such:

```
CXXFLAGS=-stdlib=libc++ CXX=clang++ meson build --default-library=static
ninja -C build
```

It fails in 1 of 2 ways, but neither of them are due to the filesystem lib not being found. Both are related to dependencies being compiled with libstdc++

1) If you have extra/fmt installed, the build fails. You can either build fmt yourself from the AUR with libc++, or let meson build fmt from the subproject.
2) The tests fail to build because gmock and gtest are linking in object archives compiled against libstdc++. The source for these is included under /usr/src/g{test,mock}, and the build system would need to be adjusted to build this under libc++.

If you resolve 1 and ignore 2, you can already compile auracle with libc++.

```
$ objdump -p build/auracle | awk '/NEEDED/ { print $2 }'
libfmt.so
libalpm.so.12
libcurl.so.4
libsystemd.so.0
libc++.so.1
libc++abi.so.1
libm.so.6
libgcc_s.so.1
libc.so.6
```

**Comment by E5ten** [comment]
You'd need to have libstdc++fs installed and it only doesn't get linked because of --as-needed.

**Comment by falconindy** [comment]
Ok, are you suggesting you have an Arch system that doesn't have libstdc++ installed?

**Comment by E5ten** [comment]
libstdc++fs.a is provided by gcc.

**Comment by E5ten** [comment]
As of gcc 9, the filesystem functions have been merged into libstdc++ and don't require libstdc++fs, the same is true for libc++ since LLVM 9. Would you consider deciding that the minimum requirement for building auracle is gcc/LLVM 9+, and not searching for or linking libstdc++fs? It would make your gcc requirement newer, but clang with libc++ would also start working without having to patch the build file to remove the stdc++fs linking.

**Comment by falconindy** [comment]
Do you mean as of gcc10?

I'm probably fine with this in principle, but it also seems like it wouldn't be a big deal to make this conditional on compiler versions (i.e. checking via cpp.get_id() and cpp.version()).

**Comment by E5ten** [comment]
I see this mentioned in the gcc 9 release notes, I'm think that's when it happened.

**Comment by falconindy** [comment]
Let's just get rid of it, then.

**Comment by E5ten** [comment]
Great, that's the last of the problems related to building with libc++, so this issue completely resolved.

**Comment by falconindy** [comment]
FWIW you still can't build the tests with libc++. This is, in part, because arch ships a precompiled libgtest.so/libgmock.so with the gtest/gmock packages.

meson has special support for gtest/gmock that's purports to [build gtest/gmock from source](https://mesonbuild.com/Dependencies.html#gtest-and-gmock). This indeed works, but the logic is such that the precompiled lib is favored over building from source and there's no way to override this behavior.

I'd be willing to accept a patch that disables the tests to work around this.
